### PR TITLE
Make history.as.bytes true by default to improve efficiency of writing MessageHistory as bytes #851

### PR DESCRIPTION
**Pull request by peter-lawrey**
Writing the `MessageHistory` as a `SelfDescribingMarshallable` showed up as significant contributor in a recent benchmark.
This makes writing the direct bytes format, the default which can be turned of using the `history.self.describing` property
Regardless of this setting, either format is read as it was written.
The estimated impact for even a short history is typically around 0.5 us, and 1.1 us on the 99.9%ile.

### CODE REVIEW

**Review by tgd — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/wire/BinaryWire.java [hunk 1]
```diff
@@ -439,6 +444,14 @@ private void copyOne(@NotNull WireOut wire, boolean first) throws InvalidMarshal
         }
     }
 
+    private void copyHistoryMessage(@NotNull WireOut wire) {
+        VanillaMessageHistory vmh = new VanillaMessageHistory();
```
Comment (tgd) [inline src/main/java/net/openhft/chronicle/wire/BinaryWire.java]: Does this need to be pooled or pulled into a field reference and recycled?

File: src/main/java/net/openhft/chronicle/wire/BinaryWire.java [hunk 2]
```diff
@@ -810,6 +831,11 @@ private StringBuilder readSpecialField(int peekCode, CharSequence keyName, int k
             case FIELD_NUMBER:
                 bytes.uncheckedReadSkipOne();
                 long fieldId = bytes.readStopBit();
+                if (fieldId == MethodReader.MESSAGE_HISTORY_METHOD_ID) {
+                    sb.setLength(0);
+                    sb.append(MethodReader.HISTORY);
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/wire/BinaryWire.java:1164]: why is this required? generated method readers handle MESSAGE_HISTORY_METHOD_ID already

File: src/test/java/net/openhft/chronicle/wire/MessageHistoryTest.java [hunk 3]
```diff
@@ -77,8 +77,8 @@ public void checkSerialiseBytes() {
         history2.readMarshallable(bw);
         assertEquals("VanillaMessageHistory { " +
                 "sources: [1=0xff,2=0xfff], " +
-                "timings: [1000000000000000000,1000000000000010000,120962203520100], " +
-                "addSourceDetails=false }", history2.toString());
+                "timings: [1000000000000000000,1000000000000010000,120962203520100,120962203520100], " +
```
Comment (JerryShea) [inline src/test/java/net/openhft/chronicle/wire/MessageHistoryTest.java:104]: why different?

File: src/test/java/net/openhft/chronicle/wire/channel/echo/RecordHistoryEchoHandlerTest.java [hunk 4]
```diff
@@ -41,6 +41,9 @@ public class RecordHistoryEchoHandlerTest extends WireTestCommon {
     private static void doTest(ChronicleContext context, ChannelHandler handler) {
         ChronicleChannel channel = context.newChannelSupplier(handler).connectionTimeoutSecs(1).get();
         assertTrue(channel.recordHistory());
+        VanillaMessageHistory history = (VanillaMessageHistory) MessageHistory.get();
+        history.reset(1, 128);
+//        history.historyMethodId(false);
```
Comment (JerryShea) [inline src/test/java/net/openhft/chronicle/wire/channel/echo/RecordHistoryEchoHandlerTest.java]: pls remove commented-out code

File: src/main/java/net/openhft/chronicle/wire/BinaryWire.java [hunk 5]
```diff
@@ -460,7 +473,13 @@ public void readWithLength(@NotNull WireOut wire, int len) throws InvalidMarshal
         try {
             bytes.readLimit(newLimit);
             @NotNull final ValueOut wireValueOut = wire.getValueOut();
-            switch (getBracketTypeNext()) {
+            BracketType bracketType = getBracketTypeNext();
+            switch (bracketType) {
+                case UNKNOWN:
+                    bytes.uncheckedReadSkipOne();
+                    copyHistoryMessage(wire);
+                    return;
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/wire/BinaryWire.java:655]: Is it right to use `UNKNOWN` to convey `HISTORY_MESSAGE` here? seems a bit obfuscated to me.

File: src/main/java/net/openhft/chronicle/wire/channel/ChronicleChannel.java [hunk 6]
```diff
@@ -88,7 +88,8 @@ default <T> T readOne(StringBuilder eventType, Class<T> expectedType) throws Clo
             try (DocumentContext dc = readingDocument()) {
                 if (dc.isPresent()) {
                     ValueIn in = dc.wire().read(eventType);
-                    if (StringUtils.isEqual(eventType, MethodReader.HISTORY)) {
+                    if (StringUtils.isEqual(eventType, MethodReader.HISTORY) ||
+                            StringUtils.isEqual(eventType, "" + MethodReader.MESSAGE_HISTORY_METHOD_ID)) {
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/wire/channel/ChronicleChannel.java:92]: I assume this gets optimised into a constant by the compiler, if not perhaps we should make it a constant?

File: src/main/java/net/openhft/chronicle/wire/BinaryWire.java (hunk 1 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/wire/BinaryWire.java]: Agree

**Review by tgd — CHANGES_REQUESTED** [review]
Hi Peter - flipping the status on this to request changes as there are a few comments

File: src/main/java/net/openhft/chronicle/wire/BinaryWire.java (hunk 5 above)
Comment (peter-lawrey) [reply src/main/java/net/openhft/chronicle/wire/BinaryWire.java:655]: The problem is that this code is widely used and the HISTORY_MESSAGE only applies here. The alternative is to create a code just for this purpose, or add an option that is ignored or produces an error in the cases where it doesn't apply.

File: src/main/java/net/openhft/chronicle/wire/channel/ChronicleChannel.java (hunk 6 above)
Comment (peter-lawrey) [reply src/main/java/net/openhft/chronicle/wire/channel/ChronicleChannel.java:92]: It should as it's known at compile time.

File: src/main/java/net/openhft/chronicle/wire/BinaryWire.java (hunk 5 above)
Comment (peter-lawrey) [reply src/main/java/net/openhft/chronicle/wire/BinaryWire.java:655]: I have added an enum as it improves clarity without as much overhead as I feared.

File: src/main/java/net/openhft/chronicle/wire/BinaryWire.java (hunk 2 above)
Comment (peter-lawrey) [reply src/main/java/net/openhft/chronicle/wire/BinaryWire.java:1164]: MethodReaders will recognise the field numbers however many tools don't use generated method readers. e.g. Chronicle Reader

File: src/test/java/net/openhft/chronicle/wire/MessageHistoryTest.java (hunk 3 above)
Comment (peter-lawrey) [reply src/test/java/net/openhft/chronicle/wire/MessageHistoryTest.java:104]: Added addSourceDetails(true); to SetTimeMessageHistory for the benefit of other tests

File: src/test/java/net/openhft/chronicle/wire/MessageHistoryTest.java [hunk 7]
```diff
@@ -77,8 +77,8 @@ public void checkSerialiseBytes() {
...
+                "addSourceDetails=true }", history2.toString());
```
Comment (peter-lawrey) [inline src/test/java/net/openhft/chronicle/wire/MessageHistoryTest.java:105]: @JerryShea printed here

File: src/test/java/net/openhft/chronicle/wire/channel/echo/RecordHistoryEchoHandlerTest.java (hunk 4 above)
Comment (peter-lawrey) [reply src/test/java/net/openhft/chronicle/wire/channel/echo/RecordHistoryEchoHandlerTest.java]: done

File: src/main/java/net/openhft/chronicle/wire/BinaryWire.java (hunk 2 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/wire/BinaryWire.java:1164]: Unless you give it `-m` argument. But point taken

Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/wire/BinaryWire.java:1164]: Could we add a comment explaining why?

File: src/main/java/net/openhft/chronicle/wire/BinaryWireCode.java [hunk 8]
```diff
@@ -35,7 +35,7 @@ public enum BinaryWireCode {
     // sequence of length 0 - 255
 //        public static final int BYTES_LENGTH64 = 0x83;
 
-    public static final int BYTES_MARSHALLABLE = 0x86; // explicitly BytesMarshallable.
+    public static final int HISTORY_MESSAGE = 0x86; // was BYTES_MARSHALLABLE, but only used for this purpose
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/wire/BinaryWireCode.java]: we should probably go back in time and mark `BYTES_MARSHALLABLE` as deprecated? Or are you confident enough that it is a internal or domestic API and won't be likely to be used outside of Chronicle products?

**Review by JerryShea — APPROVED** [review]

**Review by nicktindall — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/wire/BinaryWireCode.java (hunk 8 above)
Comment (peter-lawrey) [reply src/main/java/net/openhft/chronicle/wire/BinaryWireCode.java]: Harmless to leave as Deprecated for removal.

**Review by tgd — APPROVED** [review]

**Review by RobAustin — APPROVED** [review]
